server/service/endpoint_utils.go

```diff
@@ -107,20 +108,22 @@ func makeDecoder(iface interface{}) kithttp.DecodeRequestFunc {
 				if err != nil {
 					return nil, err
 				}
-			}
 
-			if ok && urlTagValue == "list_options" {
-				opts, err := listOptionsFromRequest(r)
-				if err != nil {
-					return nil, err
+				if urlTagValue == "list_options" {
+					opts, err := listOptionsFromRequest(r)
+					if err != nil {
+						return nil, err
+					}
+					field.Set(reflect.ValueOf(opts))
+					continue
 				}
-				field.Set(reflect.ValueOf(opts))
-				continue
-			}
 
-			if ok {
 				id, err := idFromRequest(r, urlTagValue)
-				if err != nil && err == errBadRoute && !optional {
+				if err != nil {
+					if err == errBadRoute && optional {
+						continue
+					}
+
```

Comment:
Above this is just a simplification of the conditions, and a bug I found with optional parameters.